### PR TITLE
drivers: modem: gsm_ppp: Enable hooking the setup process

### PR DESCRIPTION
**Pull request by tautologyclub**
This enables quirky modems, non-standard radio techs etc to work under
Zephyr without forcing developers to maintain huge patches. Previously,
there was no interface for:

- configuring NB-IoT or LTE CAT-M1,
- specifying the bandmask,
- disabling 3G

or a whole slew of other things that one might have to do to accomodate
modems that require slightly non-standard setup steps. This commit
should solve this problem, at least for most modems.

Signed-off-by: Benjamin Lindqvist <benjamin.lindqvist@endian.se>

### CODE REVIEW

**Comment by tautologyclub** [comment]
This requires more work, but first I'd like general input.

This at least enables us to run LTE CAT M-1 and NB-IoT on U-Blox SARA R4, and configure some roaming settings on a SIMCOM7600E, without having to maintain out-of-tree patches. This would not have been possible without this patch.

**Comment by HrMitrev** [comment]
Is there any updates on the subject? To me it looks quite interesting at least as an idea. I would even prefer if all AT commands are out in the user code and there's no CONFIG_MODEM_GSM_GENERIC / CONFIG_MODEM_GSM_QUECTEL. This way I can support multiple add-on modems based on mdm_model string without the need of different firmware. @ycsin

File: drivers/modem/Kconfig.gsm
Comment (ycsin) [inline drivers/modem/Kconfig.gsm:116]: fix formatting

File: drivers/modem/gsm_ppp.c
Comment (ycsin) [inline drivers/modem/gsm_ppp.c:677]: Would `LOG_ERR` be a better choice?

File: drivers/modem/gsm_ppp.c
Comment (ycsin) [inline drivers/modem/gsm_ppp.c:678]: The driver could potentially stuck here forever

File: drivers/modem/gsm_ppp.c
Comment (ycsin) [inline drivers/modem/gsm_ppp.c:693]: error?

File: drivers/modem/gsm_ppp.c
Comment (ycsin) [inline drivers/modem/gsm_ppp.c:761]: error?

**Review by ycsin — CHANGES_REQUESTED** [review]
I think you need to add two more states in the `gsm_finalize_connection` for the added hooks

**Comment by jukkar** [comment]
> Are there any updates on the subject? To me it looks quite interesting at least as an idea. I would even prefer if all AT commands are out in the user code and there's no CONFIG_MODEM_GSM_GENERIC / CONFIG_MODEM_GSM_QUECTEL. This way I can support multiple add-on modems based on mdm_model string without the need of different firmware.

I also like the idea that vendor specific configuration would be out of generic driver, although I am not sure that all AT commands would be in user code.  If user has a generic GSM modem without any vendor stuff, there should be no need to do anything special in application part. If the application defines any callbacks, it can override configurations but if it does not define any callbacks, then the default would be used.

**Comment by endian-benjamin** [comment]
@jukkar @ycsin @HrMitrev I am still looking for general comments on this approach (let's hold off on bickering about indentation etc), what looks iffy and what more will be required? For example, this still requires the hooks to do stuff like:

```
#include "../drivers/modem/modem_cmd_handler.h"
```

Should we perhaps make these internal headers a bit less internal somehow?

Also, the hook prototypes are pretty weird, but I believe those args are required. An alternative would be to expose the `gsm` struct. But that feels like killing mosquitos with a sledgehammer.

**Comment by ycsin** [comment]
> I am still looking for general comments on this approach

I'd probably have a function to register callback when the state of the gsm driver changes, I think this *might* be easier after the merge of the FSM implementation, that way I think user can have better control over the gsm driver.

```c
int gsm_ppp_change_state(struct gsm_modem *gsm, enum gsm_state target_state, void *user_data) {
	int ret = 0;

	gsm->state = target_state;
	if (gsm->state_change_cb) {
		ret = gsm->state_change_cb(gsm, target_state, user_data);
	}

	return ret;
}
```

We probably can have a cb for things like the registration state / rssi as well (application might be able to use these data), but I dont have any real use case for these.

**Comment by endian-benjamin** [comment]
> > I am still looking for general comments on this approach
> 
> I'd probably have a function to register callback when the state of the gsm driver changes, I think this _might_ be easier after the merge of the FSM implementation, that way I think user can have better control over the gsm driver.
> 
> ```c
> int gsm_ppp_change_state(struct gsm_modem *gsm, enum gsm_state target_state, void *user_data) {
> 	int ret = 0;
> 
> 	gsm->state = target_state;
> 	if (gsm->state_change_cb) {
> 		ret = gsm->state_change_cb(gsm, target_state, user_data);
> 	}
> 
> 	return ret;
> }
> ```
> 
> We probably can have a cb for things like the registration state / rssi as well (application might be able to use these data), but I dont have any real use case for these.

So the gsm setup state is an external API now?

**Comment by ycsin** [comment]
> So the gsm setup state is an external API now?

It could be public when necessary

**Comment by endian-benjamin** [comment]
> > So the gsm setup state is an external API now?
> 
> It could be made public when necessary

Well... I think internal state should be kept hidden as much as possible. This was discussed in another issue already I believe. External APIs need to be stable. The gsm FSM doesn't seem like a stable API to me.

**Pull request closed without merging**
